elasticapm/instrumentation/packages/graphql.py

```diff
@@ -0,0 +1,107 @@
+#  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+#  FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+#  DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+#  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+#  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+#  OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+#  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+from elasticapm import set_transaction_name
+from elasticapm.instrumentation.packages.base import AbstractInstrumentedModule
+from elasticapm.traces import capture_span
+
+
+class GraphQLExecutorInstrumentation(AbstractInstrumentedModule):
+    name = "graphql"
+
+    instrument_list = [
+        ("graphql.execution.executors.sync", "SyncExecutor.execute"),
+        ("graphql.execution.executors.gevent", "GeventExecutor.execute"),
+        ("graphql.execution.executors.asyncio", "AsyncioExecutor.execute"),
+        ("graphql.execution.executors.process", "ProcessExecutor.execute"),
+        ("graphql.execution.executors.thread", "ThreadExecutor.execute_in_thread"),
+        ("graphql.execution.executors.thread", "ThreadExecutor.execute_in_pool"),
+    ]
+
+    def call(self, module, method, wrapped, instance, args, kwargs):
+        name = "GraphQL"
+
+        info = ""
+        query = args[2]
+
+        if "ResolveInfo" == type(query).__name__:
+            if str(query.return_type) in [
+                    'Boolean',
+                    'Context',
+                    'Date',
+                    'DateTime',
+                    'Decimal',
+                    'Dynamic',
+                    'Float',
+                    'ID',
+                    'Int',
+                    'String',
+                    'Time',
+                    'UUID',
+                    'Boolean',
+                    'String'
+            ]:
+                return wrapped(*args, **kwargs)
+
+            op = query.operation.operation
+            field = query.field_name
+            info = "%s %s" % (op, field)
+        elif "RequestParams" == type(query).__name__:
+            info = "%s %s" % ("request", query.query)
+        else:
+            info = str(query)
+
+        with capture_span(
+                "%s.%s" % (name, info),
+                span_type="external",
+                span_subtype="graphql",
+                span_action="query"
+        ):
+            return wrapped(*args, **kwargs)
+
+
+class GraphQLBackendInstrumentation(AbstractInstrumentedModule):
+    name = "graphql"
+
+    instrument_list = [
+        ("graphql.backend.core", "GraphQLCoreBackend.document_from_string"),
+        ("graphql.backend.cache", "GraphQLCachedBackend.document_from_string"),
+    ]
+
+    def get_graphql_tx_name(self, graphql_doc):
+        op = graphql_doc.definitions[0].operation
```

Comment:
There is a bug here - the first item in the definition is not necessarily an operation.
It can be a fragment in which case you'll get an exception here.

@RyanKung

Reply:
(in fact, client libraries like Apollo send fragments first and operation last so this will always fail. Its best to iterate the list and find the operation then assume its first\last)

Reply:
@ekampf thanks for the heads up! Would you mind opening a new issue for this?

Reply:
@ekampf Maybe like this? Although I'm not sure how to test that....

```py
    def get_graphql_tx_name(self, graphql_doc):
        op_def = [
            i for i in graphql_doc.definitions
            if type(i).__name__=="OperationDefinition"
        ][0]
        op = op_def.operation
        fields = op_def.selection_set.selections
        return "GraphQL %s %s" % (op.upper(), "+".join([f.name.value for f in fields]))
```

Reply:
@ekampf related PR here: https://github.com/elastic/apm-agent-python/pull/881